miosix/filesystem/littlefs/lfs_miosix.cpp

```diff
@@ -45,31 +79,46 @@ int miosix::LittleFS::open(intrusive_ref_ptr<FileBase> &file, StringPart &name,
   if (mountFailed())
     return -ENOENT;
 
-  // Convert MIOSIX flags to LITTLEFS flags
-  int lfsFlags = 0;
+  // Check if we are trying to open a file or a directory
+  if (flags & _FDIRECTORY) {
+    return open_directory(file, name, flags, mode);
+  } else {
+    return open_file(file, name, flags, mode);
+  }
+}
 
-  if (flags & O_RDONLY)
-    lfsFlags |= LFS_O_RDONLY;
-  if (flags & O_WRONLY)
-    lfsFlags |= LFS_O_WRONLY;
-  if (flags & O_APPEND)
-    lfsFlags |= LFS_O_APPEND;
-  if (flags & O_CREAT)
-    lfsFlags |= LFS_O_CREAT;
-  if (flags & O_TRUNC)
-    lfsFlags |= LFS_O_TRUNC;
-  if (flags & O_EXCL)
-    lfsFlags |= LFS_O_EXCL;
+int miosix::LittleFS::open_directory(intrusive_ref_ptr<FileBase> &directory,
+                                     StringPart &name, int flags, int mode) {
+  directory = intrusive_ref_ptr<LittleFSDirectory>(
+      new LittleFSDirectory(intrusive_ref_ptr<LittleFS>(this)));
 
-  struct lfs_file littlefsFile;
-  int err = lfs_file_open(&lfs, &littlefsFile, name.c_str(), flags);
+  int err = lfs_dir_open(
+      &lfs,
+      static_cast<LittleFSDirectory *>(directory.get())->getDirReference(),
+      name.c_str());
+  if (err) {
+    directory.reset();
+    return convert_lfs_error_into_posix(err);
+  }
+
+  static_cast<LittleFSDirectory *>(directory.get())->setAsOpen();
+  return 0;
+}
+
+int miosix::LittleFS::open_file(intrusive_ref_ptr<FileBase> &file,
+                                StringPart &name, int flags, int mode) {
+  file = intrusive_ref_ptr<LittleFSFile>(
+      new LittleFSFile(intrusive_ref_ptr<LittleFS>(this), flags & O_SYNC));
 
+  int err = lfs_file_open(
+      &lfs, static_cast<LittleFSFile *>(file.get())->getFileReference(),
+      name.c_str(), convert_posix_open_to_lfs_flags(flags));
   if (err) {
+    file.reset();
     return convert_lfs_error_into_posix(err);
   }
+  static_cast<LittleFSFile *>(file.get())->setAsOpen();
 
-  file = intrusive_ref_ptr<LittleFSFile>(new LittleFSFile(
-      intrusive_ref_ptr<LittleFS>(this), littlefsFile, flags & O_SYNC));
   return 0;
 }
```

Comment:
Quite boring that I need to allocate `file`, and when failing I have to free it (`.reset()`) but I didn't think of a better solution